Fix hardware flow control on NRF52 series

marcuschangarm opened the pull request:
### Description

Due to buggy flow control logic in the UARTE, the stop signal is not being set as it is supposed to when the the module is not ready to receive data.

This commit signals the sender to halt transmitting when a DMA buffer is full and only continue again when the atomic FIFO buffer has been emptied. This allows platforms with hardware flow control to minimize all buffers and rely on flow control instead.


<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Breaking change

Commit:
Fix hardware flow control on NRF52 series

Due to buggy flow control logic in the UARTE, the stop signal
is not being set as it is supposed to when the the module is
not ready to receive data.

This commit signals the sender to halt transmitting when a DMA
buffer is full and only continue again when the atomic FIFO
buffer has been emptied. This allows platforms with hardware
flow control to minimize all buffers and rely on flow control
instead.

marcuschangarm commented:
@naveenkaje @TacoGrandeTX

naveenkaje commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/serial_api.c:
```diff
@@ -331,6 +331,20 @@ static void nordic_nrf5_uart_timeout_handler(uint32_t instance)
         /* No activity detected, no timeout set. */
         nordic_nrf5_uart_state[instance].ticker_is_running = false;
 
+        /* Check if hardware flow control is set and signal sender to stop.
```
This block of code is repeated 3 times. Would it make sense to wrap this up in it's own function?

marcuschangarm replied:
It's being called twice. It interrupt handler is convoluted as it is, I don't think adding another function will make it easier to read.

naveenkaje commented:
Please provide reference to the bug in UARTE (such as errata from Nordic) / a Github issue? It will be useful.

marcuschangarm commented:
The bug is an artifact of the UARTE API, because there is no safe way to flush the DMA buffer and swap it for a new one. When you follow the documentation and flush the DMA you'll get a ~6 ms period where the UARTE is not ready to receive. During this period the hardware flow control is inactive because the hardware flow control is only ready when the UARTE is ready to receive. 

This PR manually toggles the flow control pin during this timeout window.

TacoGrandeTX commented:
Sounds like a dicey problem! Do you have a test that showcases the problem and fix?  DMA buffers are just 1 byte - right?

marcuschangarm commented:
DMA buffers are at least 5 bytes. You can connect a logic analyzer to any application. You'll notice the RTS line is never set even though it should be.

deepikabhavnani commented:
Build requested by @marcuschangarm 
Query : Is it for 5.10.1 or 5.101.2?  

/morph build

0xc0170 commented:
/morph export-build

theotherjimmy commented:
/morph test

NRF52_DK timed out.

theotherjimmy commented:
@marcuschangarm Is there a chance that the NRF52 timeout could have been caused by this change?

NirSonnenschein commented:
failures in tests for NRF52_DK  timeouts and sync issues. this doesn't seem incidental. @marcuschangarm please take a look

marcuschangarm commented:
/morph test

cmonr commented:
/morph build

New commit > new SHA > new Build step needed.

cmonr commented:
@marcuschangarm Please take a look at the test failures. Failures appear to be related to PR

cmonr commented:
Huh. Looks like CI had some issues allocating devices to use for Tests, and timed out as a result.

Rerunning.

/morph test

studavekar commented:
Internal ticket IOTSYSTOOL-1197created before we re-trigger

studavekar commented:
Tools are updated to handle faulty sync, triggered to check if we have better results

/morph test

marcuschangarm commented:
/morph build

cmonr commented:
Oh, huh. Looks like the auto-run tests resolved themselves.

@marcuschangarm https://github.com/ARMmbed/mbed-os/pull/8292#issuecomment-429321078
_Tsk tsk tsk_

NirSonnenschein commented:
@marcuschangarm failures in test look PR related, please take a look.

marcuschangarm commented:
@studavekar @NirSonnenschein 
The failures occur in a test with heavy UART traffic. I suspect the CI infrastructure not following the flow control signal causes the Tx buffer to overflow.

studavekar commented:
> @studavekar @NirSonnenschein
> The failures occur in a test with heavy UART traffic. I suspect the CI infrastructure not following the flow control signal causes the Tx buffer to overflow.

@marcuschangarm not sure if Daplink supports flow control from host <-->daplink<---- hw fw ctrl ---> device.

We also need to make review the test, we should avoid heavy uart traffic in tests. If we want to stress uart it should part of a separate test.

Failure : 
1) ```[1539363423.79][CONN][RXD] :144::FAIL: Expected 'passed' Was 'pased'``` - **passed** --> **pased**

cmonr commented:
@c1728p9 
> not sure if Daplink supports flow control from host <-->daplink<---- hw fw ctrl ---> device.

Thoughts?

c1728p9 commented:
> not sure if Daplink supports flow control from host <-->daplink<---- hw fw ctrl ---> device.

I took a look at the DAPLink code and flow control is enabled for the NRF52 boards.

marcuschangarm commented:
@c1728p9 
Does DAPLink support flow control on the host side as well?

c1728p9 commented:
@marcuschangarm yep, it is built into the USB CDC protocol and can't be turned off.

marcuschangarm commented:
@c1728p9 
So if the CI enables flow control then it should work all the way down?

marcuschangarm commented:
The failing test is for a component that is not supported by the NRF52_DK target.

Both the `rtc_time` and `rtc_time_conv` HAL tests are missing this header:

```
#if !DEVICE_RTC
#error [NOT_SUPPORTED] RTC API not supported for this target
#endif
```

0xc0170 commented:
@mprse Can you review ^^ ?

c1728p9 commented:
> So if the CI enables flow control then it should work all the way down?

Yep

mprse commented:
> Both the rtc_time and rtc_time_conv HAL tests are missing this header:
> 
> > #if !DEVICE_RTC
> > #error [NOT_SUPPORTED] RTC API not supported for this target
> > #endif
 
These tests verifies time conversion functions from the following [library](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_mktime.h) which is available for all targets and does not require `RTC`. Also even when `RTC` is not supported on the specific platform it is then replaced by `LPTICKER` if available (which is `NRF52_DK` case I think):
https://github.com/ARMmbed/mbed-os/blob/140f3e20d6b7ddc4ab32229761362f754d078761/platform/mbed_rtc_time.cpp#L25-L74

I can confirm that the failing test sends lots of data between host and device using uart. So this is a great test for this PR 😅 . Looking at the last results we have failure on `IAR` and `ARM` compilers.
For `IAR` there is a regular test timeout case, so we could consider increasing the test execution timeout. 
For `ARM` we got some kind of communication error:
`:144::FAIL: Expected 'passed' Was 'pased'`
Looks like one byte has been lost during the transmission which is bad and may cause some unexpected issues in the future, so I think this needs to be investigated.

marcuschangarm commented:
So, what's the plan?
* Characters are missing, but the CI doesn't support flow control.
* Tests are failing for components we don't claim support for.

Commit:
Remove debug flow control from NRF52_DK and NRF52840_DK

Currently flow control is not supported by the CI and enabling it
causes unwanted side effects.

marcuschangarm commented:
/morph build

0xc0170 commented:
Eclipse error, but all good there

/morph export-build
